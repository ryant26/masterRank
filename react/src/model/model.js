import {
    addHero as addHeroAction,
    removeHero as removeHeroAction,
    clearAllHeroes as clearAllHeroesAction
} from "../actionCreators/hero";
import {
    addHero as addPreferredHeroAction,
    removeHero as removePreferredHeroAction,
    updateHeroes as updatePreferredHeroesAction
} from "../actionCreators/preferredHeroes";
import { updateUser as updateUserAction } from "../actionCreators/user";
import {
    addFilter as addFilterAction,
    removeFilter as removeFilterAction,
    removeAllFilters as removeAllFiltersAction
} from "../actionCreators/heroFilters";
import {
    updateGroup as updateGroupAction,
    leaveGroup as leaveGroupAction
} from '../actionCreators/group';
import { clientEvents } from "../api/websocket";
import {
    addGroupInvite as addGroupInviteAction,
    removeGroupInvite as removeGroupInviteAction
} from '../actionCreators/groupInvites';
import {
    pushBlockingEvent as pushBlockingLoadingAction,
    popBlockingEvent as popBlockingLoadingAction,
} from "../actionCreators/loading";

import {
    joinedGroupNotification,
    userJoinedGroupNotification,
    inviteSentNotification,
    inviteReceivedNotification,
    successfullyLeftGroupNotification,
    errorNotification
} from '../components/Notifications/Notifications';

import NotRealHeroes from '../resources/metaListFillerHeroes';

let socket;
let store;

const initialize = function(passedSocket, passedStore) {
    store = passedStore;
    socket = passedSocket;

    store.dispatch(pushBlockingLoadingAction());

    socket.on(clientEvents.initialData, (heroesFromServer) => _handleInitialData(heroesFromServer));
    socket.on(clientEvents.heroAdded, (hero) => _addHeroToStore(hero));
    socket.on(clientEvents.heroRemoved, (hero) => _removeHeroFromStore(hero));

    socket.on(clientEvents.groupInviteReceived, (groupInviteObject) => _addGroupInviteToStore(groupInviteObject));
    socket.on(clientEvents.playerInvited, (groupInviteObject) => _updateGroupInStore(groupInviteObject));

    socket.on(clientEvents.groupInviteCanceled, (groupInviteObject) => _removeGroupInviteFromStore(groupInviteObject));
    socket.on(clientEvents.playerInviteCanceled, (groupInviteObject) => _updateGroupInStore(groupInviteObject));

    socket.on(clientEvents.groupInviteDeclined, (groupInviteObject) => _updateGroupInStore(groupInviteObject));
    socket.on(clientEvents.groupInviteAccepted, (groupInviteObject) => _handleGroupInviteAccepted(groupInviteObject));
    socket.on(clientEvents.groupPromotedLeader, (groupInviteObject) => _updateGroupInStore(groupInviteObject));
    socket.on(clientEvents.playerHeroLeft, (groupInviteObject) => _updateGroupInStore(groupInviteObject));

    socket.on(clientEvents.error.addHero, _addHeroErrorHandler);
    socket.on(clientEvents.error.groupLeave, _groupErrorHandler);
    socket.on(clientEvents.error.groupInviteAccept, _groupErrorHandler);
    socket.on(clientEvents.error.groupInviteCancel, _groupErrorHandler);
    socket.on(clientEvents.error.groupInviteDecline, _groupErrorHandler);
};

const addHeroFilterToStore = function(filter) {
    store.dispatch(addFilterAction(filter));
};

const removeHeroFilterFromStore = function(filter) {
    store.dispatch(removeFilterAction(filter));
};

const removeAllHeroFiltersFromStore = function() {
    store.dispatch(removeAllFiltersAction());
};

const addPreferredHeroToStore = function(heroName, preference) {
    store.dispatch(addPreferredHeroAction(heroName, preference));
};

const removePreferredHeroFromStore = function(heroName, preference) {
    store.dispatch(removePreferredHeroAction(heroName, preference));
};

const updatePreferredHeroesInStore = function(heroes) {
    store.dispatch(updatePreferredHeroesAction(heroes));
};


const updatePreferredHeroes = function(heroes) {
    let currentPreferredHeroes = store.getState().preferredHeroes.heroes;
    let numberOfHeroesToCheck = Math.max(heroes.length, currentPreferredHeroes.length);

    for (let i = 0; i < numberOfHeroesToCheck; i++) {
        let currentPreferredHero = store.getState().preferredHeroes.heroes[i];
        let newPreferredHero = heroes[i];

        if (currentPreferredHero !== newPreferredHero){
            if (currentPreferredHero) {
                socket.removeHero(currentPreferredHero);
            }

            if (newPreferredHero) {
                socket.addHero(newPreferredHero, i+1);
            }
        }
    }

    updatePreferredHeroesInStore(heroes);
};

const updateUser = function(user) {
    store.dispatch(updateUserAction(user));
};

const inviteUserToGroup = function(userObject) {
    inviteSentNotification(userObject.platformDisplayName);
    socket.groupInviteSend(userObject);
};

const createNewGroup = function() {
    socket.createGroup(store.getState().preferredHeroes.heroes[0]);
};

const leaveGroup = function() {
    successfullyLeftGroupNotification(store.getState().group.leader.platformDisplayName);
    store.dispatch(leaveGroupAction());
    socket.groupLeave();
};

const cancelInvite = function(userObject) {
    socket.groupInviteCancel(userObject);
};

const acceptGroupInviteAndRemoveFromStore = function(groupInviteObject) {
    _removeGroupInviteFromStore(groupInviteObject);
    joinedGroupNotification(groupInviteObject.leader.platformDisplayName);
    socket.groupInviteAccept(groupInviteObject.groupId);
};

const declineGroupInviteAndRemoveFromStore = function(groupInviteObject) {
    _removeGroupInviteFromStore(groupInviteObject);
    socket.groupInviteDecline(groupInviteObject.groupId);
};

const loadMetaListFillerHeroes = () => {
    NotRealHeroes.forEach((hero) => {
        _addHeroToStore(hero);
    });
};

const _handleInitialData = function(heroesFromServer) {
    store.dispatch(clearAllHeroesAction());
    loadMetaListFillerHeroes();

    let userPlatformDisplayName = store.getState().user.platformDisplayName;
    heroesFromServer.forEach((hero) => {
        if(hero.platformDisplayName !== userPlatformDisplayName){
            store.dispatch(addHeroAction(hero));
        } else  {
            socket.removeHero(hero.heroName);
        }
    });

    store.getState().preferredHeroes.heroes.forEach((heroName, i) => {
         socket.addHero(heroName, (i+1));
    });

    store.dispatch(popBlockingLoadingAction());
};

const _addHeroToStore = function(hero) {
    store.dispatch(addHeroAction(hero));
    if (hero.platformDisplayName === store.getState().user.platformDisplayName) {
        addPreferredHeroToStore(hero.heroName, hero.preference);
    }

    if (store.getState().preferredHeroes.heroes[0] === hero.heroName) {
        createNewGroup(store.getState().preferredHeroes.heroes[0]);
    }
};

const _removeHeroFromStore = function(hero) {
    store.dispatch(removeHeroAction(hero));
    //TODO: Multiple tabs or window messes up preferred heroes. 
    if (hero.platformDisplayName === store.getState().user.platformDisplayName) {
        removePreferredHeroFromStore(hero.heroName, hero.priority);
    }
};

const _addHeroErrorHandler = function(error) {
    errorNotification(error.heroName);
    removePreferredHeroFromStore(error.heroName);
    // Do whatever we do to shuffle heroes in the case that multiple were added
    // After this one
};

const _addGroupInviteToStore = function(groupInviteObject) {
    inviteReceivedNotification(groupInviteObject.leader.platformDisplayName);
    store.dispatch(addGroupInviteAction(groupInviteObject));
};

const _removeGroupInviteFromStore = function(groupInviteObject) {
    store.dispatch(removeGroupInviteAction(groupInviteObject));
};

const _updateGroupInStore = function(groupInviteObject) {
    store.dispatch(updateGroupAction(groupInviteObject));
};

const _handleGroupInviteAccepted = (newGroup) => {
    //TODO: is there a better way to get the new group member?
    let previouslyPendingMembers = store.getState().group.pending;

    if (previouslyPendingMembers.length > 0 && newGroup.members.length > 0) {
        let newMember = previouslyPendingMembers.find((pendingMember) => {
            return newGroup.members.find((member) => {
                return member.platformDisplayName === pendingMember.platformDisplayName;
            });
        });
        //TODO: this if statement should never not trigger. If it does trigger that means something went wrong.
        // Should i sent a generic user joined group message or just not send a message?
        if(newMember) {
            userJoinedGroupNotification(newMember.platformDisplayName);
        }
    }

    store.dispatch(updateGroupAction(newGroup));
};

const _groupErrorHandler = (error) => {
    errorNotification(error.message);
};

const Actions = {
    initialize,
    addPreferredHeroToStore,
    updatePreferredHeroes,
    addHeroFilterToStore,
    removeHeroFilterFromStore,
    removeAllHeroFiltersFromStore,
    updateUser,
    leaveGroup,
    inviteUserToGroup,
    createNewGroup,
    cancelInvite,
    acceptGroupInviteAndRemoveFromStore,
    declineGroupInviteAndRemoveFromStore
};

export default Actions;