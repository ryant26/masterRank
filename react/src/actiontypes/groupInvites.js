export const ADD_GROUP_INVITE = 'groupInvites/ADD_GROUP_INVITE';