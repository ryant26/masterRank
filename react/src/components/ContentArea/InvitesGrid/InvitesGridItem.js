import React from 'react';
import PropTypes from 'prop-types';
import HeroCard from '../DashboardHome/HeroRoles/HeroCard/HeroCard';

const InvitesGridItem = ({invite}) => {
    let numberOfMembers = invite.members.length + 1;
    let baseSr = invite.leader.skillRating;
    baseSr = invite.members.reduce((sr, member) => sr + member.skillRating, baseSr);
    let avgSrString = (baseSr/numberOfMembers).toFixed(0);

    let groupHeroes = [<HeroCard key={invite.leader.platformDisplayName} user={invite.leader} hero={invite.leader}/>];
    invite.members.forEach((member) => {
        groupHeroes.push(<HeroCard key={member.platformDisplayName} user={member} hero={member}/>);
    });

    return (
        <div className="InvitesGridItem flex flex-column card">
            <div className="header">
                <div className="group-title flex align-center justify-center">
                    <h3>{invite.leader.platformDisplayName}'s</h3>
                    <h3 className="regular">Group</h3>
                </div>
                <div className="groupSR sub-title flex align-center justify-center">
                    {avgSrString} Group SR
                </div>
            </div>
            <div className="body flex flex-column">{groupHeroes}</div>
            <div className="button-group flex justify-around">
                <div className="button button-five flex justify-center align-center">
                    <div className="button-content">
                        DECLINE
                    </div>
                </div>
                <div className="button button-secondary flex justify-center align-center">
                    <div className="button-content">
                        ACCEPT
                    </div>
                </div>
            </div>
        </div>
    );
};

InvitesGridItem.propTypes = {
    invite: PropTypes.object.isRequired
};

export default InvitesGridItem;