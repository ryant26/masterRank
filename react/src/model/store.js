import { createStore as createReduxStore } from 'redux';

import reducers from '../reducers/index';
import { loadState } from './reduxSerializer';


export const createStore = function() {
    store = createReduxStore(reducers, loadState());

    return store;
};

let store = createStore();

export default store;