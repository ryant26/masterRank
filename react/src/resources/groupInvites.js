const groupInvites = [
  {
    groupId: 10,
    groupSize: 4,
    inviteDate: "2018-02-19T12:57:24.790Z",
    leader: {
      platformDisplayName: "luckbomb#1234",
      skillRating: 3200,
      heroName: "genji",
      stats: {}
    },
    members: [
      {
        platformDisplayName: "wismo0#1234",
        skillRating: 2200,
        heroName: "widowmaker",
        stats: {}
      }
    ],
    pending: [
      {
        platformDisplayName: "andyLin#1234",
        skillRating: 1800,
        heroName: "tracer",
        stats: {}
      }
    ]
  },
  {
    groupId: 8,
    groupSize: 3,
    inviteDate: "2017-09-19T21:57:24.790Z",
    leader: {
      platformDisplayName: "luckbomb#1234",
      skillRating: 3200,
      heroName: "genji",
      stats: {}
    },
    members: [],
    pending: [
      {
        platformDisplayName: "cutie#1320",
        skillRating: 1600,
        heroName: "tracer",
        stats: {}
      }
    ]
  },
    {
        groupId: 2,
        groupSize: 4,
        inviteDate: "2017-06-19T16:57:24.790Z",
        leader: {
            platformDisplayName: "luckbomb#1234",
            skillRating: 3200,
            heroName: "doomfist",
            stats: {}
        },
        members: [
            {
                platformDisplayName: "wismo0#1234",
                skillRating: 2200,
                heroName: "soldier76",
                stats: {}
            }
        ],
        pending: [
            {
                platformDisplayName: "cutie#1320",
                skillrating: 1600,
                heroName: "winston",
                stats: {}
            }
        ]
    },
     {
      groupId: 210,
      groupSize: 6,
      inviteDate: "2018-02-19T16:57:24.790Z",
      leader: {
        platformDisplayName: "luckbomb#1234",
        skillRating: 3200,
        heroName: "genji",
        stats: {}
      },
      members: [
        {
          platformDisplayName: "wismo0#1234",
          skillRating: 2200,
          heroName: "widowmaker",
          stats: {}
        },{
          platformDisplayName: "andyLin#1234",
          skillRating: 1600,
          heroName: "winston",
          stats: {}
        }
      ],
      pending: [
        {
          platformDisplayName: "cutie#1320",
          skillRating: 1800,
          heroName: "tracer",
          stats: {}
        },{
          platformDisplayName: "pwnShopp#4810",
          skillRating: 1800,
          heroName: "soldier76",
          stats: {}
        },{
          platformDisplayName: "banannaTruck#1872",
          skillRating: 1800,
          heroName: "doomfist",
          stats: {}
        }
      ]
    },
];

export default groupInvites;