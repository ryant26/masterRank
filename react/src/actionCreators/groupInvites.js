import * as GroupInviteActionTypes from 'actiontypes/groupInvites';

export const addGroupInvite = (invite) => {
  return {
    type: GroupInviteActionTypes.ADD_GROUP_INVITE,
    invite
  };
};

export const removeGroupInvite = (invite) => {
  return {
    type: GroupInviteActionTypes.REMOVE_GROUP_INVITE,
    invite
  };
};