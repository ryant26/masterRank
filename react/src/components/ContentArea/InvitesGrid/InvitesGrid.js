import React from 'react';
import {Link} from 'react-router-dom';
import {home} from 'components/Routes/links';
import {connect} from 'react-redux';
import Badge from 'components/Badge/Badge';
import InvitesGridItem from './InvitesGridItem';
import PropTypes from 'prop-types';

const InvitesGrid = ({groupInvites}) => {
    const gridItems = groupInvites.map((invite) => <InvitesGridItem key={invite.groupId} invite={invite}/>);

    return (
        <div className="InvitesGrid flex flex-column grow">
            <Link className="homeLink" to={home} style={{ textDecoration: 'none' }}>
                <div className="actionable-text"><i className="fa fa-long-arrow-left"/> back to players list</div>
            </Link>
            <div className="title flex align-center">
                <h2>Group Invites Pending</h2>
                <Badge number={groupInvites.length}/>
            </div>
            <div className="grid flex wrap">
                {gridItems}
            </div>
        </div>
    );
};

InvitesGrid.propTypes = {
    groupInvites: PropTypes.array.isRequired
};

const mapStateToProps = (state) => {
    return {
        groupInvites: state.groupInvites
    };
};

export default connect(mapStateToProps)(InvitesGrid);

