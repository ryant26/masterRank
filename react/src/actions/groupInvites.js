import * as GroupInviteActionTypes from '../actiontypes/groupInvites';

export const addGroupInvite = (invite) => {
  return {
    type: GroupInviteActionTypes.ADD_GROUP_INVITE,
    invite
  };
};