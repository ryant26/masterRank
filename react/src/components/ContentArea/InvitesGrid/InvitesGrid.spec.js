import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import InvitesGrid from './InvitesGrid';
import {createStore} from "../../../model/store";
import groupInvites from '../../../resources/groupInvites';
import {addGroupInvite} from "../../../actions/groupInvites";
import {MemoryRouter} from 'react-router-dom';

const getStore = () => {
    let store = createStore();
    groupInvites.forEach((invite) => {
        store.dispatch(addGroupInvite(invite));
    });
    return store;
};

const mountWithStore = (store=getStore()) => {
    return mount(
        <MemoryRouter>
            <InvitesGrid store={store}/>
        </MemoryRouter>
    );
};

const renderWithStore = (store=getStore()) => {
    return renderer.create(
        <MemoryRouter>
            <InvitesGrid store={store}/>
        </MemoryRouter>
    );
};


describe('InvitesGrid Component', () => {
    it('should render without exploding', () => {
        const wrapper = mountWithStore();
        const InvitesGridComponent = wrapper.find(InvitesGrid);
        expect(InvitesGridComponent).toBeTruthy();
    });

    it('should match the snapshot', () => {
        const component = renderWithStore();
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});