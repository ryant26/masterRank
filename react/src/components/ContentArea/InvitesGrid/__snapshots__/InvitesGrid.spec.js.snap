// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvitesGrid Component should match the snapshot 1`] = `
<div
  className="InvitesGrid flex flex-column grow"
>
  <a
    className="homeLink"
    href="/"
    onClick={[Function]}
    style={
      Object {
        "textDecoration": "none",
      }
    }
  >
    <div
      className="actionable-text"
    >
      <i
        className="fa fa-long-arrow-left"
      />
       back to players list
    </div>
  </a>
  <div
    className="title flex align-center"
  >
    <h2>
      Group Invites Pending
    </h2>
    <div
      className="Badge"
    >
      <div
        className="background flex justify-center align-center"
      >
        <div
          className="label"
        >
          1
        </div>
      </div>
    </div>
  </div>
  <div
    className="grid flex wrap"
  >
    <div
      className="InvitesGridItem flex flex-column card"
    >
      <div
        className="header"
      >
        <div
          className="group-title flex align-center justify-center"
        >
          <h3>
            luckbomb#1234
            's
          </h3>
          <h3
            className="regular"
          >
            Group
          </h3>
        </div>
        <div
          className="groupSR sub-title flex align-center justify-center"
        >
          2700
           Group SR
        </div>
      </div>
      <div
        className="body flex flex-column"
      >
        <div
          className="HeroCard flex align-center"
        >
          <div
            className="invitePlayerButton"
          >
            <img
              alt="genji icon"
              className="HeroImage"
              onClick={undefined}
              src={
                Object {
                  "process": [Function],
                }
              }
            />
            <div
              className="overlay"
            >
              <div
                className="plus-container flex align-center justify-center"
                onClick={[Function]}
              >
                <i
                  className="fa fa-plus"
                />
              </div>
            </div>
          </div>
          <div
            className="content flex flex-column"
          >
            <div
              className="display-name"
            >
              luckbomb#1234
            </div>
            <div
              className="sub-title"
            >
              <span>
                3200
                 SR
              </span>
              <span>
                 | 
              </span>
              <span>
                0
                % WR
              </span>
            </div>
          </div>
          <div
            className="button-primary"
            onClick={[Function]}
          >
            <div
              className="button-content"
            >
              stats
            </div>
          </div>
        </div>
        <div
          className="HeroCard flex align-center"
        >
          <div
            className="invitePlayerButton"
          >
            <img
              alt="widowmaker icon"
              className="HeroImage"
              onClick={undefined}
              src={
                Object {
                  "process": [Function],
                }
              }
            />
            <div
              className="overlay"
            >
              <div
                className="plus-container flex align-center justify-center"
                onClick={[Function]}
              >
                <i
                  className="fa fa-plus"
                />
              </div>
            </div>
          </div>
          <div
            className="content flex flex-column"
          >
            <div
              className="display-name"
            >
              wismo0#1234
            </div>
            <div
              className="sub-title"
            >
              <span>
                2200
                 SR
              </span>
              <span>
                 | 
              </span>
              <span>
                0
                % WR
              </span>
            </div>
          </div>
          <div
            className="button-primary"
            onClick={[Function]}
          >
            <div
              className="button-content"
            >
              stats
            </div>
          </div>
        </div>
      </div>
      <div
        className="button-group flex justify-around"
      >
        <div
          className="button button-six flex justify-center align-center"
          onClick={[Function]}
        >
          <div
            className="button-content"
          >
            DECLINE
          </div>
        </div>
        <div
          className="button button-secondary flex justify-center align-center"
          onClick={[Function]}
        >
          <div
            className="button-content"
          >
            ACCEPT
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
